ako-gateway-api/k8s/validator.go

```diff
@@ -318,3 +322,119 @@ func IsHTTPRouteConfigValid(key string, obj *gatewayv1.HTTPRoute) bool {
+				if listener.TLS != nil {
+					if (listener.TLS.Mode != nil && *listener.TLS.Mode != gatewayv1.TLSModeTerminate) || len(listener.TLS.CertificateRefs) == 0 {
+						continue
+					}
+					for _, certRef := range listener.TLS.CertificateRefs {
+						// TODO: In future, ref grant impact needs to be considered.
+						certNamespace := namespace
+						if certRef.Namespace != nil {
+							certNamespace = string(*certRef.Namespace)
+						}
+						// add condition for checking gateway status
+						if gwStatus != nil && string(certRef.Name) == name && certNamespace == namespace {
+							setConditions(gwStatus, listenerIndex, gatewayObj.ObjectMeta.Generation, deleteFlag)
+							certFound = true
+							break
+						}
+					}
+				}
+				if gwStatus != nil && !akogatewayapilib.IsListenerInvalid(gwStatus, listenerIndex) {
+					validListenerCount += 1
+				}
+			} // listener loop ends
+			// if cert not found, that means it is stale entry in secret to gateway map
+			if certFound {
+				setGatewayCondition(gwStatus, validListenerCount, len(gatewayObj.Spec.Listeners), gatewayObj.ObjectMeta.Generation)
+			} else {
+				//remove the entry from the clone
+				clonedGWList = utils.Remove(clonedGWList, gwNSName)
+			}
+		} // gw loop ends
+		// update the mapping
+		akogatewayapiobjects.GatewayApiLister().UpdateSecretToGateway(secretNSName, clonedGWList)
+	}
+}
+
+// Mapping has to be taken care between secret and gateway
+func setConditions(gwStatus *gatewayv1.GatewayStatus, index int, generation int64, isDelete bool) {
+	if !isDelete {
+		akogatewayapistatus.NewCondition().
+			Type(string(gatewayv1.ListenerConditionAccepted)).
+			Reason(string(gatewayv1.ListenerReasonAccepted)).
+			Message("Listener is valid").
+			Status(metav1.ConditionTrue).
+			ObservedGeneration(generation).SetIn(&gwStatus.Listeners[index].Conditions)
+
+		akogatewayapistatus.NewCondition().
+			Type(string(gatewayv1.ListenerConditionResolvedRefs)).
+			Reason(string(gatewayv1.ListenerReasonResolvedRefs)).
+			Message("Reference is valid").
+			Status(metav1.ConditionTrue).
+			ObservedGeneration(generation).SetIn(&gwStatus.Listeners[index].Conditions)
+		return
+	}
+	akogatewayapistatus.NewCondition().
+		Type(string(gatewayv1.ListenerConditionAccepted)).
+		Reason(string(gatewayv1.ListenerReasonInvalid)).
+		Message("Listener is Invalid").
+		Status(metav1.ConditionFalse).
+		ObservedGeneration(generation).SetIn(&gwStatus.Listeners[index].Conditions)
+
+	akogatewayapistatus.NewCondition().
+		Type(string(gatewayv1.ListenerConditionResolvedRefs)).
+		Reason(string(gatewayv1.ListenerReasonInvalidCertificateRef)).
+		Message("Secret does not exist").
+		Status(metav1.ConditionFalse).
+		ObservedGeneration(generation).SetIn(&gwStatus.Listeners[index].Conditions)
+}
+
+func setGatewayCondition(gwStatus *gatewayv1.GatewayStatus, validListenerCount, totalListenerCount int, observedGeneration int64) {
+	if validListenerCount == 0 {
```

Comment:
can be restructured as if{.....}elif{.....}else{....}

Reply:
this should be fine